File: Core/Stateflows.Common/Events/CompoundRequest.cs
using System.Linq;
using System.Collections.Generic;
using Newtonsoft.Json;
using Stateflows.Common.Utilities;

namespace Stateflows.Common
{
    public interface ICompoundRequestBuilder
    {
        ICompoundRequestBuilder Add<TEvent>(TEvent @event, IEnumerable<EventHeader> headers = null);
        ICompoundRequestBuilder AddRange<TEvent>(params TEvent[] events);
    }

    public class CompoundRequest : IRequest<CompoundResponse>, ICompoundRequestBuilder
    {
        [JsonProperty(TypeNameHandling = TypeNameHandling.None)]
        public List<EventHolder> Events { get; set; } = [];

        ICompoundRequestBuilder ICompoundRequestBuilder.Add<TEvent>(TEvent @event, IEnumerable<EventHeader> headers)
            => Add(@event, headers);

        public CompoundRequest Add<TEvent>(TEvent @event, IEnumerable<EventHeader> headers = null)
        {
            Events.Add(@event.ToTypedEventHolder(headers));

            return this;
        }

        ICompoundRequestBuilder ICompoundRequestBuilder.AddRange<TEvent>(params TEvent[] events)
            => AddRange(events);

        public CompoundRequest AddRange<TEvent>(params TEvent[] events)
        {
            Events.AddRange(events.Select(@event => @event.ToTypedEventHolder()));

            return this;
        }
    }
}

File: Transport/@stateflows/common/src/events/compound.request.ts
import { Request } from "./request";
import { CompoundResponse } from "./compound.response";
import { EventHolder } from "../classes/event-holder";
import { StateflowsEvent } from "../decorators/stateflows-event";

@StateflowsEvent("Stateflows.Common.CompoundRequest, Stateflows.Common")
export class CompoundRequest extends Request<CompoundResponse> {
    constructor(
        events: Array<any>,
    ) {
        super();
        this.events = events.map(event => new EventHolder(event));
    }

    public events: Array<EventHolder>;
}
